docs/architecture/adr-042-state-sync.md

```diff
@@ -0,0 +1,209 @@
+(ie. elements aren't indexed by the element number).
+
+This design could be generalized to have numbered chunks instead of
+elements, where chunks are indexed by their chunk number and determined
+by some algorithm for grouping the elements in the tree. This would
+allow pre-computation and persistence of the chunks that could alleviate
+the concerns about real-time iteration and random reads.  However, there
+seems to be a larger issue with the "numbered chunks" approach: it does
+not seem possible, in general, to verify that the chunk corresponds to
+the chunk number, even if it can be verified that the chunk is a valid
+set of nodes in the tree. A malicious peer could provide a chunk that
+contains valid nodes from the tree but does not actually match the
+alleged chunk number. This could make it difficult for nodes to figure
+out which parts of the tree they have and which parts they still need.
+Hence, to go with this chunking approach to state sync, information
+about how state is chunked would need to be hashed into the state
+itself, or into the block header, so that chunk numbers could be
+directly verified.
+
+An alternative design was suggested by Jae Kwon in
+[#3639](https://github.com/tendermint/tendermint/issues/3639) where chunking
+happens lazily and in a dynamic way: nodes request key ranges from their peers,
+and peers respond with some subset of the
+requested range and with notes on how to request the rest in parallel from other
+peers. Unlike chunk numbers, keys can be verified directly. And if some keys in the
+range are ommitted, proofs for the range will fail to verify.
+This way a node can start by requesting the entire tree from one peer,
+and that peer can respond with say the first few keys, and the ranges to request
+from other peers.
+
+#### Eager StateSync
+Warp Sync as implemented in Parity
+["Warp Sync"](https://wiki.parity.io/Warp-Sync-Snapshot-Format.html) to rapidly
+download both blocks and state snapshots from peers. Data is carved into ~4MB
+chunks and snappy compressed. Hashes of snappy compressed chunks are stored in a
+manifest file which co-ordinates the state-sync. Obtaining a correct manifest
+file seems to require an honest majority of peers. This means you may not find
+out the state is incorrect until you download the whole thing and compare it
+with a verified block header.
+
+A similar solution was implemented by Binance in
+[#3594](https://github.com/tendermint/tendermint/pull/3594)
+based on their initial implementation in
+[PR #3243](https://github.com/tendermint/tendermint/pull/3243)
+and [some learnings](https://docs.google.com/document/d/1npGTAa1qxe8EQZ1wG0a0Sip9t5oX2vYZNUDwr_LVRR4/edit).
+Note this still requires the honest majority peer assumption.
+One major advantage of the warp-sync approach is that chunks can be compressed before sending.
+
+Ideally State Sync in Tendermint would not require an honest majority of peers, and instead
```

Comment:
Maybe it's worth being more precise here and say that in public systems we normally don't make assumptions about ratio of faulty processes among node peers. This might be reasonable assumption in enterprise blockchain systems, although even in that case I would challenge that assumption. For example, what if system tolerates 10 faulty processes (f=10, n=31), and you have most of those processes in node peer set (for example peer set size is also 10).